### PR TITLE
feat(CLI): allow for custom plugins

### PR DESCRIPTION
**Pull request by willemneal**
### What

Now any binary with the prefix `soroban-`, e.g. soroban-plugin will be executed with `soroban plugin`. Similar to `cargo-`.

Also `soroban --list` will list the installed plugins.

### Why

Allows devs to create plugins. E.g. `soroban network` could be used for launching local nodes.

### Known limitations

[TODO or N/A]

### CODE REVIEW

File: cmd/soroban-cli/src/commands/custom.rs [hunk 1]
```diff
@@ -0,0 +1,29 @@
+use std::process::Command;
+use which::which;
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Plugin not provided. Should be `soroban plugin` for a binary `soroban-plugin`")]
+    MissingSubcommand,
+    #[error(transparent)]
+    IO(#[from] std::io::Error),
+    #[error("Cannot find executable: {0}")]
```
Comment (paulbellamy) [inline cmd/soroban-cli/src/commands/custom.rs:10]: Something like the cargo errors here might be nice:
```
$ cargo whatever
error: no such subcommand: `whatever`

     View all installed subcommands with `cargo --list`
```

Then `cargo --list` gives you a nice list of all the known subcommands.

**Review by paulbellamy — COMMENTED** [review]
Looks great so far, a couple questions.

File: cmd/soroban-cli/src/bin/main.rs [hunk 2]
```diff
@@ -1,11 +1,18 @@
...
+use clap::{CommandFactory, Parser};
+use soroban_cli::{commands::custom::run_external, Root};
 
 #[tokio::main]
 async fn main() {
-    let root = Root::new().unwrap_or_else(|e| {
-        let mut cmd = Root::command();
-        e.format(&mut cmd).exit();
+    let root = Root::try_parse().unwrap_or_else(|e| {
+        if let clap::error::ErrorKind::UnknownArgument | clap::error::ErrorKind::InvalidSubcommand =
```
[the remark this answers is not in the document]
Comment (willemneal) [reply cmd/soroban-cli/src/bin/main.rs:7]: Yep!

File: cmd/soroban-cli/src/commands/custom.rs (hunk 1 above)
Comment (willemneal) [reply cmd/soroban-cli/src/commands/custom.rs]: Done! Even added a suggestion for closely matched words.

File: cmd/soroban-cli/src/commands/custom.rs
[the remark this answers is not in the document]
Comment (willemneal) [reply cmd/soroban-cli/src/commands/custom.rs:1]: Done.

File: cmd/soroban-cli/src/bin/main.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (willemneal) [reply cmd/soroban-cli/src/bin/main.rs]: fixed

File: cmd/crates/soroban-test/tests/it/plugin.rs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+    // Call soroban with the PATH variable set to include the target/bin directory
+    assert_cmd::Command::cargo_bin("soroban")
+        .unwrap()
+        .arg("hello")
+        .env("PATH", &paths)
+        .assert()
+        .stdout("Hello, world!\n");
+
+    // Call soroban with the PATH variable set to include the target/bin directory
```
Comment (willemneal) [inline cmd/crates/soroban-test/tests/it/plugin.rs:28]:
```suggestion
    // Call `soroban --list` with the PATH variable set to include the target/bin directory
```

File: cmd/soroban-cli/src/bin/main.rs [hunk 4]
```diff
@@ -1,11 +1,23 @@
...
+use clap::{CommandFactory, Parser};
+use soroban_cli::{commands::plugin, Root};
 
 #[tokio::main]
 async fn main() {
-    let root = Root::new().unwrap_or_else(|e| {
-        let mut cmd = Root::command();
-        e.format(&mut cmd).exit();
+    let root = Root::try_parse().unwrap_or_else(|e| {
+        if std::env::args().any(|s| &s == "--list") {
```
Comment (paulbellamy) [inline cmd/soroban-cli/src/bin/main.rs:7]: Suggestion: This seems like it ends up handling `--list` anywhere in the command:
e.g. `soroban contract --list`

Which is a bit surprising. Shouldn't this only be hit if there is no subcommand?

Comment (paulbellamy) [reply cmd/soroban-cli/src/bin/main.rs:7]: Like:
```diff
diff --git a/cmd/soroban-cli/src/bin/main.rs b/cmd/soroban-cli/src/bin/main.rs
index ea07f92..be9d2e1 100644
--- a/cmd/soroban-cli/src/bin/main.rs
+++ b/cmd/soroban-cli/src/bin/main.rs
@@ -4,10 +4,6 @@ use soroban_cli::{commands::plugin, Root};
 #[tokio::main]
 async fn main() {
     let root = Root::try_parse().unwrap_or_else(|e| {
-        if std::env::args().any(|s| &s == "--list") {
-            println!("{}", plugin::list().unwrap_or_default().join("\n"));
-            std::process::exit(0);
-        }
         if let clap::error::ErrorKind::InvalidSubcommand = e.kind() {
             if let Err(error) = plugin::run() {
                 eprintln!("error: {error}");
diff --git a/cmd/soroban-cli/src/commands/mod.rs b/cmd/soroban-cli/src/commands/mod.rs
index 829e82b..ee42c35 100644
--- a/cmd/soroban-cli/src/commands/mod.rs
+++ b/cmd/soroban-cli/src/commands/mod.rs
@@ -54,7 +54,7 @@ pub struct Root {
     #[arg(long)]
     pub list: bool,
     #[command(subcommand)]
-    pub cmd: Cmd,
+    pub cmd: Option<Cmd>,
 }
 
 impl Root {
@@ -72,12 +72,20 @@ impl Root {
     }
     pub async fn run(&self) -> Result<(), Error> {
         match &self.cmd {
-            Cmd::Contract(contract) => contract.run().await?,
-            Cmd::Events(events) => events.run().await?,
-            Cmd::Lab(lab) => lab.run().await?,
-            Cmd::Version(version) => version.run(),
-            Cmd::Completion(completion) => completion.run(),
-            Cmd::Config(config) => config.run()?,
+            Some(Cmd::Contract(contract)) => contract.run().await?,
+            Some(Cmd::Events(events)) => events.run().await?,
+            Some(Cmd::Lab(lab)) => lab.run().await?,
+            Some(Cmd::Version(version)) => version.run(),
+            Some(Cmd::Completion(completion)) => completion.run(),
+            Some(Cmd::Config(config)) => config.run()?,
+            None => {
+                if self.list {
+                    println!("{}", plugin::list().unwrap_or_default().join("\n"));
+                } else {
+                    let _ = Self::command().print_help();
+                    std::process::exit(1);
+                }
+            }
         };
         Ok(())
     }
```

File: cmd/soroban-cli/src/commands/plugin.rs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    };
+    let re = regex::Regex::new(re_str).unwrap();
+    Ok(which::which_re(re)?
+        .filter_map(|b| {
+            let s = b.file_name()?.to_str()?;
+            Some(s.strip_suffix(".exe").unwrap_or(s).to_string())
+        })
+        .filter(|s| !(is_hex_string(s) && s.len() > MAX_HEX_LENGTH))
+        .map(|s| s.replace("soroban-", ""))
+        .collect())
```
Comment (paulbellamy) [inline cmd/soroban-cli/src/commands/plugin.rs:74]: Seems like there's a bug in the hex processing on windows, because the test is failing

**Review by paulbellamy — COMMENTED** [review]
Almost there, I think. One suggestion, one failing test on windows...

File: cmd/soroban-cli/src/bin/main.rs (hunk 4 above)
Comment (willemneal) [reply cmd/soroban-cli/src/bin/main.rs:7]: Fixed it by checking if there is a missing subcommand.  Pushed, but still need to figure out the windows hex string issue.

File: cmd/soroban-cli/src/commands/plugin.rs (hunk 5 above)
Comment (willemneal) [reply cmd/soroban-cli/src/commands/plugin.rs:74]: Yeah I need to check on this with a local windows test.

**Comment by willemneal** [comment]
@paulbellamy I fixed the issue you raised with `--list` working anywhere.  Now it's just if there is a missing subcommand.

As for the windows issue, I changed the test to just check for the `hello` plugin.  I figure it's better to get the plugins in now and figure out the issue of an extra listed plugin in a future PR.

**Review by paulbellamy — APPROVED** [review]

**Pull request merged**